specs/lazy-adr/adr-021-transaction-streaming.md

```diff
@@ -0,0 +1,154 @@
+# ADR 021: Transaction Streaming via Mini Blocks from Single Sequencer
+
+## Changelog
+
+- 2025-04-28: Initial draft
+
+## Context
+
+Currently, in single sequencer mode in Rollkit, every configured block time, the sequencer produces blocks. This includes constructing a batch of transactions to put in a block, executing them, and producing a header with an updated state root. This process introduces latency between transaction submission and network-wide propagation, as transactions are only shared after full block production. This ADR proposes a design where the sequencer streams smaller transaction batches in the form of "mini blocks" sooner to full nodes via p2p so full nodes can receive transactions without waiting for execution or state root computation by the sequencer to reduce overall end-user latency and enhance UX.
+
+### Current Architecture (Before)
+```mermaid
+flowchart LR
+    subgraph "Block Time (2s)"
+        Txs[Transactions] --> Seq[Sequencer]
+        Seq --> |Execute Txs| Block[Block]
+        Block --> |Compute State Root| Nodes[Full Nodes]
+    end
+    
+    style Seq fill:#f9f,stroke:#333,stroke-width:2px,color:#000
+    style Block fill:#bbf,stroke:#333,stroke-width:2px,color:#000
+    style Nodes fill:#bfb,stroke:#333,stroke-width:2px,color:#000
+    style Txs color:#000
+```
+
+### Proposed Architecture (After)
+```mermaid
+flowchart LR
+    subgraph "Mini-Block Time (200ms)"
+        Txs[Transactions] --> Seq[Sequencer]
+        Seq --> MB1[Mini-Block 1]
+        Seq --> MB2[Mini-Block 2]
+        Seq --> MB3[Mini-Block n]
+        MB1 & MB2 & MB3 --> |Stream| Nodes[Full Nodes]
+    end
+    
+    subgraph "Block Time (2s)"
+        MB1 & MB2 & MB3 --> Block[Block]
+        Block --> |Compute State Root| Nodes
+    end
+    
+    style Seq fill:#f9f,stroke:#333,stroke-width:2px,color:#000
+    style MB1 fill:#ddf,stroke:#333,stroke-width:2px,color:#000
+    style MB2 fill:#ddf,stroke:#333,stroke-width:2px,color:#000
+    style MB3 fill:#ddf,stroke:#333,stroke-width:2px,color:#000
+    style Block fill:#bbf,stroke:#333,stroke-width:2px,color:#000
+    style Nodes fill:#bfb,stroke:#333,stroke-width:2px,color:#000
+    style Txs color:#000
+```
+
+## Alternative Approaches
+
+1. **Immediate Transaction Broadcasting**
+   - Broadcast individual transactions immediately upon receipt
+   - Pros: Lowest possible latency
+   - Cons: Higher network overhead, potential complex coordination with execution layer
+
+2. **Parallel Block Production**
+   - Run block production in parallel while streaming transactions
+   - Pros: Maintains traditional block structure
+   - Cons: More complex implementation, potential consistency issues on handling transactions with state contention
+
+3. **Variable Block Times**
```

Comment:
we should mention this is already present with lazy mode. Also, would this system conflict with lazy in anyway? when would a user use this vs lazy?